File: src-tauri/src/models/torrent.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TorrentAddedResponse {
    pub id: usize,
    pub name: String,
    pub info_hash: String,
    pub files: Vec<TorrentFileInfo>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TorrentSummary {
    pub id: usize,
    pub name: String,
    pub info_hash: String,
    pub state: TorrentState,
    pub progress: f64,
    pub download_speed: u64,
    pub upload_speed: u64,
    pub peers_connected: usize,
    pub total_bytes: u64,
    pub downloaded_bytes: u64,
    pub file_count: usize,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TorrentDetails {
    pub id: usize,
    pub name: String,
    pub info_hash: String,
    pub state: TorrentState,
    pub progress: f64,
    pub download_speed: u64,
    pub upload_speed: u64,
    pub peers_connected: usize,
    pub total_bytes: u64,
    pub downloaded_bytes: u64,
    pub file_count: usize,
    pub files: Vec<TorrentFileInfo>,
    pub output_folder: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TorrentFileInfo {
    pub index: usize,
    pub name: String,
    pub path: String,
    pub length: u64,
    pub is_playable: bool,
    pub mime_type: Option<String>,
    pub stream_url: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TorrentAddOptions {
    pub output_folder: Option<String>,
    pub only_files: Option<Vec<usize>>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum TorrentState {
    Initializing,
    Downloading,
    Paused,
    Completed,
    Error,
}

File: src/lib/types/torrent.ts
export interface TorrentAddedResponse {
  id: number;
  name: string;
  info_hash: string;
  files: TorrentFileInfo[];
}

export interface TorrentSummary {
  id: number;
  name: string;
  info_hash: string;
  state: TorrentState;
  progress: number;
  download_speed: number;
  upload_speed: number;
  peers_connected: number;
  queued_peers: number;
  connecting_peers: number;
  total_bytes: number;
  downloaded_bytes: number;
  uploaded_bytes?: number;
  file_count: number;
}

export interface TorrentDetails extends TorrentSummary {
  files: TorrentFileInfo[];
  added_at: string;
  output_folder: string;
}

export interface TorrentFileInfo {
  index: number;
  name: string;
  path: string;
  length: number;
  is_playable: boolean;
  mime_type: string | null;
  stream_url: string | null;
}

export interface TorrentAddOptions {
  output_folder?: string;
  only_files?: number[];
}

export interface TorrentProgress {
  id: number;
  progress: number;
  download_speed: number;
  upload_speed: number;
  peers_connected: number;
  queued_peers: number;
  connecting_peers: number;
  downloaded_bytes: number;
  uploaded_bytes: number;
  total_bytes: number;
  state: TorrentState;
}

export type TorrentState =
  | "initializing"
  | "downloading"
  | "paused"
  | "completed"
  | "error";
